README.md
# fluent-plugin-amqp

This plugin provides both a Source and Matcher which uses RabbitMQ as its transport.

[![Build Status](https://travis-ci.org/giraffi/fluent-plugin-amqp.svg?branch=master)](https://travis-ci.org/giraffi/fluent-plugin-amqp)
[![Gem Version](https://badge.fury.io/rb/fluent-plugin-amqp.svg)](https://badge.fury.io/rb/fluent-plugin-amqp)
[![Coverage Status](https://coveralls.io/repos/github/giraffi/fluent-plugin-amqp/badge.svg)](https://coveralls.io/github/giraffi/fluent-plugin-amqp)

# Table of contents

1. [Requirements](#requirements)
1. [Features](#features)
    1. [Highly Available Failover](#feat-failover)
1. [Configuration](#configuration)
    1. [Common parameters](#conf-common)
    1. [Source](#conf-source)
    1. [Matcher](#conf-matcher)
         1. [Message Headers](#conf-matcher-header)
1. [Example Use Cases](#usecases)
    1. [Using AMQP instead of Fluent TCP forwarders](#uc-forwarder)
    1. [Enable TLS Authentication](#uc-tls)
1. [Contributing](#contributing)
1. [Copyright](#copyright)

# Requirements <a name="requirements"></a>


|fluent-amqp-plugin|fluent|ruby|
|----|----|----|
|>= 0.10.0 | >= 0.14.8 | >= 2.1 |
| < 0.10.0 | > 0.10.0, < 2 <sup>*</sup> | >= 1.9  |

* May not support all future fluentd features

# Features <a name="features"></a>

## Highly Available Failover <a name="feat-failover"></a>

You can use the `hosts` parameter to provide an array of rabbitmq hosts which
are in your cluster. This allows for highly avaliable configurations where a
node in your cluster may become inaccessible and this plugin will attempt a reconnection
on another host in the array.

> *WARNING:* Due to limitations in the library being used for connecting to RabbitMQ
each node of the cluster must use the same port, vhost and other configuration.

### Example

```
<source>
  type amqp
  hosts ["amqp01.example.com","amqp02.example.com"]
  port 5672
  vhost /
  user guest
  pass guest
  queue logs
  format json
</source>
```



# Configuration <a name="configuration"></a>

## A note on routing keys

If you would like to filter events from certain sources, you can make use of the
`key`, `tag_key` and `tag_header` configuration options.

The RabbitMQ [routing key](http://www.rabbitmq.com/tutorials/tutorial-four-ruby.html)
that is set for the message on the broker determines what you may be able to
filter against when consuming messages.

For example, if you want a 'catch-all' consumer that gets all messages from a
direct exchange, you should set `tag_key true` on both `source` and `matcher`. This
will then recreate the original event's tag ready for processing by the consumers
matchers.

If you want to have selective control over the messages that are consumed, you
can set `tag_key true` on the matcher, but `key some.tag` on the source. Only
messages with the given tag will be consumed, however its recommended that you
understand the difference between the different exchange types, and how multiple
consumers may impact message delivery.

## Common parameters <a name="conf-common"></a>

The following parameters are common to both matcher and source
plugins, and can be used as required.

|param|type|default|description|
|----|----|----|---|
|:host|:string|nil| *Required (if hosts unset)* Hostname of RabbitMQ server |
|:hosts|:array|nil| *Required (if host unset)* An array of hostnames of RabbitMQ server in a common cluster (takes precidence over `host`)|
|:user|:string|"guest"| Username to connect |
|:pass|:string|"guest"| Password to authenticate with (Secret) |
|:vhost|:string|"/"| RabbitMQ Virtual Host|
|:port|:integer|5672| RabbitMQ listening port|
|:durable|:bool|false| Should the queue or exchange be durable? |
|:passive|:bool|false| If true, will fail if queue or exchange does not exist |
|:auto_delete|:bool|false| Should the queue be deleted when all consumers have closed? |
|:heartbeat|:integer|60| Frequency of heartbeats to ensure quiet connections are kept open|
|:ssl|:bool|false| Is SSL enabled for this connection to RabbitMQ|
|:verify_ssl|:bool|false| Verify the SSL certificate presented by RabbitMQ |
|:tls|:bool|false| Should TLS be used for authentication |
|:tls_cert|:string|nil| *Required if `tls true`* Path (or content) of TLS Certificate |
|:tls_key|:string|nil| *Required if `tls true`* Path (or content) of TLS Key |
|:tls_ca_certificates|:array|nil| Array of paths to CA certificates |
|:tls_verify_peer|:bool|true| Verify the servers TLS certificate |
|:tag_key|:bool|false| Should the routing key be used for the event tag |
|:tag_header|:string|nil| What header should be used for the event tag |
|:time_header|:string|nil| What header should be used for the events timestamp |


## Source - Obtain events from a RabbitMQ queue <a name="conf-source"></a>


Using the amqp as a source allows you to read messages from RabbitMQ and handle
them in the same manner as a locally generated event.

It can be used in isolation; reading (well formed) events generated by other
applications and published onto a queue, or used with the amqp matcher, which
can replace the use of the fluent forwarders.


### Source specific parameters

Note: The following are in addition to the common parameters shown above.

|param|type|default|description|
|----|----|----|---|
|:tag|:string|"hunter.amqp"| Accepted events are tagged with this string (See also tag_key)|
|:queue|:string|nil| What queue contains the events to read |
|:exclusive|:bool|false| Should we have exclusive use of the queue? |
|:payload_format|:string|"json"| Deprecated - Use `format`|
|:bind_exchange|:boolean|false| Should the queue automatically bind to the exchange |
|:exchange|:string|nil| What exchange should the queue bind to? |
|:exchange_type|:string|"direct"| Type of exchange ( direct, fanout, topic, headers, x-consistent-hash, x-modulus-hash )|
|:routing_key|:string|nil| What exchange should the queue bind to? |

### Example

```
<source>
  type amqp
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  queue logs
  format json
</source>
```

## Matcher - output events from RabbitMQ <a name="conf-matcher"></a>

### Matcher specific parameters

|param|type|default|description|
|----|----|----|----|
|:exchange|:string|""| Name of the exchange to send events to |
|:exchange_type|:string|"direct"| Type of exchange ( direct, fanout, topic, headers, x-consistent-hash, x-modulus-hash )|
|:persistent|:bool|false| | Are messages kept on the exchange even if RabbitMQ shuts down |
|:key|:string|nil| Routing key to attach to events (Only applies when `exchange_type topic`) See also `tag_key`|
|:content_type|:string|"application/octet"| Content-type header to send with message |
|:content_encoding|:string|nil| Content-Encoding header to send - eg base64 or rot13 |

#### Headers <a name="conf-matcher-headers"></a>

It is possible to specify message headers based on the content of the incoming
message, or as a fixed default value as shown below;

```
<matcher ...>
...

  <header>
    name LogLevel
    source level
    default "INFO"
  </header>
  <header>
    name SourceHost
    default my.example.com
  </header>
  <header>
    name CorrelationID
    source x-request-id
  </header>
  <header>
    name NestedExample
    source a.nested.value
  </header>
  <header>
    name AnotherNestedExample
    source ["a", "nested", "value"]
  </header>

...
</matcher>
```


The header elements may be set multiple times for multiple additional headers
to be included on any given message.

* If source is omitted, the header will _always_ be set to the default value
* If default is omitted the header will only be set if the source is found
* Overloading headers is permitted
    * Last defined header with a discovered or default value will be used
    * Defaults and discovered values are treated equally - If you set a default
    for a overloaded header the earlier headers *will never be used*


### Example

```
<match **.**>
  type amqp
  key my_routing_key
  exchange amq.direct
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  content_type application/json
</match>
```

# Example Use Cases <a name="usecases"></a>

## Using AMQP instead of Fluent TCP forwarders <a name="uc-forwarder"></a>

One particular use case of the AMQP plugin is as an alternative to the built-in
fluent forwarders.

You can simply setup each client to output events to a RabbitMQ exchange
which is then consumed by one or more input agents.

The example configuration below shows how to setup a direct exchange, with
multiple consumers each receiving events.

### Matcher - Writes to Exchange

```
<match **>
  type amqp
  exchange amq.direct
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  format json
  tag_key true
</match>
```

### Source - Reads from queues

```
<source>
  type amqp
  host amqp.example.com
  port 5672
  vhost /
  user guest
  pass guest
  queue my_queue
  format json
  tag_key true
</source>
```

## Enable TLS Authentication <a name="uc-tls"></a>

The example below shows how you can configure TLS authentication using signed encryption keys
which will be validated by your appropriately configured RabbitMQ installation.

For more information on setting up TLS encryption, see the [Bunny TLS documentation](http://rubybunny.info/articles/tls.html)

Note: The 'source' configuration accepts the same arguments.

```
<match **.**>
  type amqp
  key my_routing_key
  exchange amq.direct
  host amqp.example.com
  port 5671              # Note that your port may change for TLS auth
  vhost /
  user guest
  pass guest

  tls true
  tls_key "/etc/fluent/ssl/client.key.pem"
  tls_cert "/etc/fluent/ssl/client.crt.pem"
  tls_ca_certificates ["/etc/fluent/ssl/server.cacrt.pem", "/another/ca/cert.file"]
  tls_verify_peer true

</match>
```

## Docker Container

A docker container is included in this project to help with testing and debugging.

You can simply build the docker container's ready for use with the following;
```
docker-compose build
```

Start the cluster of three containers with;
```
docker-compose up
```

And finally, submit test events, one a second, to the built in tcp.socket source
with;

```
while [ true ] ; do echo "{ \"test\": \"$(date)\" }" | nc ${DOCKER_IP} 20001; sleep 1; done
```
## Rabbitmq-sharding

You may find that rabbitmq doesn't behave nicely when delivering lots of events to a single queue as the process thread gets overloaded and starts to send flow control events back to publishers. If you're in this situation, try the rabbitmq-sharding plugin which is in RMQ 3.6+ and can allow queues to be dynamically generated per-node.

To use this;

1. Enable the plugin on all nodes `rabbitmq-plugins enable rabbitmq_sharding`
1. Create an exchange to accept events and to be sharded using `x-modulus-hash` or `x-consistent-hash`
1. Configure a sharding policy on the input exchange
    - `rabbitmqctl set_policy images-shard "^fluent.modhash$" '{"shards-per-node": 2, "routing-key": "1234"}'`
1. Setup fluentd to use the associated type and bind to a queue named the same as the input exchange name
    - This queue is created 'dynamically' and will not show as a formal queue in the manager, but will deliver events to fluent normally
    
*Warning*: You will need to run at least N consumers for the N shards created as the plugin does not try to route all shards onto consumers dynamically.

# Contributing to fluent-plugin-amqp <a name="contributing"></a>

- Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
- Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
- Fork the project
- Start a feature/bugfix branch
- Commit and push until you are happy with your contribution
- Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
- Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

# Copyright <a name="copyright"></a>

Copyright (c) 2011 Hiromi Ishii. See LICENSE.txt for
Copyright (c) 2013- github/giraffi. See LICENSE.txt for
further details.


CHANGELOG.md

## 0.12.x

## 0.12.0 - Nov 02, 2018

* giraffi/fluent-plugin-amqp#48 - Added support for Multi-Workers

## 0.11.x

## 0.11.0 - March 10, 2017

* Added `content_type` and `content_encoding` configuration for
   the Matcher plugin - #37
* Added header options to set message headers based on the message content on
  the Matcher plugin - #35
* Enabled support for Coveralls code-coverage reporting

# 0.10.x

## 0.10.1 - 2016-11-17

* BugFix: Failed to set ruby_version properly and broke 0.10.0

## 0.10.0 - 2016-11-17

* *Breaking Change* - fluent-amqp-plugin only compatible with fluent >= 0.14.8 and ruby >= 2.1
* Feature: Updated to use new FluentD 0.14 plugin format
* Feature: Support for nanosecond precision

# 0.9.x

## 0.9.3

* BugFix: `:tls => false` hangs connections for some reason #32 HT: @mrkurt

## 0.9.2

* Better error handling in AMQP Matcher to deal with byte range errors, and
any other failure which would prevent buffers from being replayed effectivly.

## 0.9.1

* BugFix: giraffi/fluent-plugin-amqp#25 - Wrapped JSON.dump with simple catch which logs error but does not attempt to fix encoding errors

## 0.9.0

* Added support for Fluent 0.14.x
    * travis-ci build validates plugin against fluent 0.10, 0.12 and 0.14
* Compatibility: Use old json library when building for Ruby 1.9 and Fluent <= 0.12

# 0.8.x

## 0.8.2

* Miscs: *no effect* Remove fixed date from gemspec #22

## 0.8.1

* Fix bug: TypeError with single RabbitMQ. #21

## 0.8.0

* Add automatic failover support for multiple hosts on same cluster. #19

# 0.7.x

## 0.7.1

* Sources can now automatically bind to exchanges [Issue #16](https://github.com/giraffi/fluent-plugin-amqp/issues/16)
* Refactored Docker container to use docker-compose and use multiple containers for testing source and matches concurrently

## 0.7.0

* TLS configuration supported [#15](https://github.com/giraffi/fluent-plugin-amqp/pull/15)
* Docker container created for testing [#13](https://github.com/giraffi/fluent-plugin-amqp/pull/13)

# 0.6.x

## 0.6.1

* Heartbeat added to outbound connections [#11](https://github.com/giraffi/fluent-plugin-amqp/pull/11)

## 0.6.0

*Breaking Change:* Ruby 2.0+ now required

* Updated dependencies for Bunny and Fluent to latest
* Changed default Heartbeat from 0 to 60 seconds

# < 0.5.x

* See commit history for breakdown of changes


test/helper.rb
require 'rubygems'
require 'bundler'
require 'simplecov'
SimpleCov.start

begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end

require 'test/unit'
require 'shoulda'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
#require 'fluent-plugin-amqp'


test/plugin/test_in_amqp.rb

require_relative '../helper'
require 'fluent/test'
require 'fluent/test/driver/input'
require 'fluent/plugin/in_amqp'

class AMPQInputTest < Test::Unit::TestCase
  def setup
    Fluent::Test.setup
  end


  CONFIG = %[
    type amqp
    format json
    host amqp.example.com
    port 5672
    vhost /
    user guest
    pass guest
    queue logs
  ]

  TLS_CONFIG = CONFIG + %[
    tls true
    tls_key "/etc/fluent/ssl/client.key.pem"
    tls_cert "/etc/fluent/ssl/client.crt.pem"
    tls_ca_certificates ["/etc/fluent/ssl/server.cacrt.pem", "/another/ca/cert.file"]
    tls_verify_peer true
  ]

  def create_driver(conf)
    Fluent::Test::Driver::Input.new(Fluent::Plugin::AMQPInput).configure(conf)
  end

  sub_test_case 'configuration' do
    test 'basic configuration' do
      configs = {'basic' => CONFIG}
      configs.merge!('tls' => TLS_CONFIG)

      configs.each_pair { |k, v|
        d = create_driver(v)
        assert_equal "amqp.example.com", d.instance.host
        assert_equal 5672, d.instance.port
        assert_equal "guest", d.instance.user
        assert_equal "/", d.instance.vhost
      }
    end

    test 'array of hosts' do
      conf = CONFIG + %[
        type amqp
        format json
        hosts ["bob","fred"]
        ]

      d = create_driver(conf)
      assert_equal ["bob", "fred"], d.instance.hosts
      assert_equal 5672, d.instance.port
      assert_equal "guest", d.instance.user
      assert_equal "/", d.instance.vhost
    end

    test 'invalid tls configuration' do
      assert_raise_message(/'tls_key' and 'tls_cert' must be all specified if tls is enabled./) do
        create_driver(CONFIG + %[tls true])
      end
    end

    test 'invalid host / queue configuration' do
      assert_raise_message(/'host\(s\)' and 'queue' must be all specified./) do
        create_driver(%[
          type amqp
          format json
          host bob
          ])
      end
    end
  end
  sub_test_case 'connection handling' do

    test 'queue is created' do
      omit("Need to replace bunny-mock with proper mocking - it cant cope with q.subscribe and sensible replay/verification")
      plugin = create_driver(CONFIG).instance
      plugin.connection = BunnyMock.new

      # Start the driver and wait while it initialises the threads etc
      plugin.start
      10.times { sleep 0.05 }

      # Should have created the 'logs' queue
      assert_equal true, plugin.connection.queue_exists?('logs')
    end

  end
end


test/plugin/test_out_amqp.rb
# encoding: utf-8

require_relative '../helper'
require 'fluent/test'
require 'fluent/test/driver/output'
require 'fluent/plugin/out_amqp'

require 'bunny-mock'

class AMPQOutputTest < Test::Unit::TestCase

  attr_reader :driver

  def setup
    Fluent::Test.setup
    BunnyMock.use_bunny_queue_pop_api = true
  end

def get_plugin(configuration = CONFIG)
  omit("BunnyMock is not avaliable") unless Object.const_defined?("BunnyMock")

  @driver = create_driver(configuration)
  plugin = @driver.instance
  plugin.connection = BunnyMock.new

  # Start the driver and wait while it initialises the threads etc
  plugin.start
  10.times { sleep 0.05 }
  return plugin
end

  CONFIG = %[
    type amqp
    format json
    host amqp.example.com
    port 5672
    vhost /
    user guest
    pass guest
    exchange my_exchange
    exchange_type fanout
    tag_key true
  ]

  TLS_CONFIG = CONFIG + %[
    tls true
    tls_key "/etc/fluent/ssl/client.key.pem"
    tls_cert "/etc/fluent/ssl/client.crt.pem"
    tls_ca_certificates ["/etc/fluent/ssl/server.cacrt.pem", "/another/ca/cert.file"]
    tls_verify_peer true
  ]

  def create_driver(conf)
    Fluent::Test::Driver::Output.new(Fluent::Plugin::AMQPOutput).configure(conf)
  end

  sub_test_case 'configuration' do
    test 'basic configuration' do
      configs = {'basic' => CONFIG}
      configs.merge!('tls' => TLS_CONFIG)

      configs.each_pair { |k, v|
        d = create_driver(v)
        assert_equal "amqp.example.com", d.instance.host
        assert_equal 5672, d.instance.port
        assert_equal "guest", d.instance.user
        assert_equal "/", d.instance.vhost
      }
    end

    test 'array of hosts' do
      conf = CONFIG + %[
        type amqp
        format json
        hosts ["bob","fred"]
        ]

      d = create_driver(conf)
      assert_equal ["bob", "fred"], d.instance.hosts
      assert_equal 5672, d.instance.port
      assert_equal "guest", d.instance.user
      assert_equal "/", d.instance.vhost
    end

    test 'invalid tls configuration' do
      assert_raise_message(/'tls_key' and 'tls_cert' must be all specified if tls is enabled./) do
        create_driver(CONFIG + %[tls true])
      end
    end

    test 'invalid host configuration' do
      assert_raise_message(/'host' or 'hosts' must be specified./) do
        create_driver(%[
          type amqp
          format json
          tag_key true
          ])
      end
    end

    test 'invalid tag / tag_key configuration' do
      assert_raise_message(/Either 'key' or 'tag_key' must be set./) do
        create_driver(%[
          type amqp
          format json
          host anywhere.example.com
          ])
      end
    end
  end

  sub_test_case 'connection handling' do

    test 'Exchange is created and bound to' do

      plugin = get_plugin()

      # Should have created the 'logs' queue
      assert_equal true, plugin.connection.exchange_exists?('my_exchange')
      pend("Unable to check binding state - cant reach channel instance") {
        chnl = plugin.somehow_get_channel
        asset_equal true, chnl.bound_to?('my_exchange')
      }
    end

  end

  sub_test_case 'message_writing' do


    test 'A simple object can be written to the broker' do
# Testing these two bits;
#      data = JSON.dump( data ) unless data.is_a?( String )
#      @exch.publish(data, :key => routing_key( tag ), :persistent => @persistent, :headers => headers( tag, time ))

      plugin = get_plugin()

      # Should have created the 'logs' queue
      assert_equal true, plugin.connection.exchange_exists?('my_exchange')

      # bind a testing queue to the exchange
      queue = plugin.channel.queue 'my.test.queue'
      queue.bind plugin.exch, routing_key: 'test'
      # queue.test is now bound to the configured exchange

      # v0.14 test driver does not permit to specify String object into #feed args.
      es = Fluent::OneEventStream.new(Time.now.to_i, 'This is a simple string')
      # Emit an event through the plugins driver
      @driver.run(default_tag: 'test') do
        @driver.feed(es)
      end

      # Validate the message was delivered
      assert_equal 1, queue.message_count
      deliveryProps, msgProps, message = queue.pop
      assert_equal 'This is a simple string', message
      assert_equal 'test', msgProps[:key]

    end

    test 'An object can be written to the broker and is converted to json' do
  # Testing these two bits;
  #      data = JSON.dump( data ) unless data.is_a?( String )
  #      @exch.publish(data, :key => routing_key( tag ), :persistent => @persistent, :headers => headers( tag, time ))

      plugin = get_plugin()

      # Should have created the 'logs' queue
      assert_equal true, plugin.connection.exchange_exists?('my_exchange')

      # bind a testing queue to the exchange
      queue = plugin.channel.queue 'my.test.queue'
      queue.bind plugin.exch, routing_key: 'test'
      # queue.test is now bound to the configured exchange

      # Emit an event through the plugins driver
      object = { message: 'This is an event', nested: { type: 'hash', value: 'banana'} }
      @driver.run(default_tag: 'test') do
        @driver.feed( object )
      end

      # Validate the message was delivered
      assert_equal 1, queue.message_count
      deliveryProps, msgProps, message = queue.pop
      assert_equal JSON.dump(object) , message

    end


    test 'Test UTF8 unicode and emoji strings do not crash the plugin' do
  # Testing these two bits;
  #      data = JSON.dump( data ) unless data.is_a?( String )
  #      @exch.publish(data, :key => routing_key( tag ), :persistent => @persistent, :headers => headers( tag, time ))

      plugin = get_plugin()

      # Should have created the 'logs' queue
      assert_equal true, plugin.connection.exchange_exists?('my_exchange')

      # bind a testing queue to the exchange
      queue = plugin.channel.queue 'my.test.queue'
      queue.bind plugin.exch, routing_key: 'test'
      # queue.test is now bound to the configured exchange

      complex_string_msg = { message: '日 🕵 iPhone\xAE \u{1f60e}' }
      # Emit an event through the plugins driver
      @driver.run(default_tag: 'test') do
        @driver.feed( complex_string_msg )
      end

      # Validate the message was _not_ delivered
      assert_equal 1, queue.message_count
      deliveryProps, msgProps, message = queue.pop
      assert_equal JSON.dump(complex_string_msg) , message

    end


      test 'Can explicitly specific content_type and encoding through configuration' do
  # Testing these two bits;
  #      data = JSON.dump( data ) unless data.is_a?( String )
  #      @exch.publish(data, :key => routing_key( tag ), :persistent => @persistent, :headers => headers( tag, time ))

        plugin = get_plugin( CONFIG + %[
          content_type application/json
          content_encoding base64 ] )

        # Should have created the 'logs' queue
        assert_equal true, plugin.connection.exchange_exists?('my_exchange')

        # bind a testing queue to the exchange
        queue = plugin.channel.queue 'my.test.queue'
        queue.bind plugin.exch, routing_key: 'test'
        # queue.test is now bound to the configured exchange

        # v0.14 test driver does not permit to specify String object into #feed args.
        # Message = Base64('{ "test": "this is a string" }')
        es = Fluent::OneEventStream.new(Time.now.to_i, 'eyAidGVzdCI6ICJ0aGlzIGlzIGEgc3RyaW5nIiB9')
        # Emit an event through the plugins driver
        @driver.run(default_tag: 'test') do
          @driver.feed(es)
        end

        # Validate the message was delivered
        assert_equal 1, queue.message_count
        deliveryProps, msgProps, message = queue.pop

        assert_equal 'eyAidGVzdCI6ICJ0aGlzIGlzIGEgc3RyaW5nIiB9', message
        assert_equal 'application/json', msgProps[:content_type]
        assert_equal 'base64', msgProps[:content_encoding]
      end
  end
end


lib/fluent/plugin/in_amqp.rb
require 'time'
require 'fluent/plugin/input'
require 'fluent/plugin/parser'
require 'bunny'

module Fluent::Plugin
  ##
  # AMQPInput to be used as a Fluent SOURCE, reading messages from a RabbitMQ
  # message broker
  class AMQPInput < Input
    Fluent::Plugin.register_input('amqp', self)

    helpers :compat_parameters, :parser

    # Bunny connection handle
    #   - Allows mocking for test purposes
    attr_accessor :connection

    config_param :tag, :string, default: "hunter.amqp"

    config_param :host, :string, default: nil
    config_param :hosts, :array, default: nil
    config_param :user, :string, default: "guest"
    config_param :pass, :string, default: "guest", secret: true
    config_param :vhost, :string, default: "/"
    config_param :port, :integer, default: 5672
    config_param :ssl, :bool, default: false
    config_param :verify_ssl, :bool, default: false
    config_param :heartbeat, :integer, default: 60
    config_param :queue, :string, default: nil
    config_param :durable, :bool, default: false
    config_param :exclusive, :bool, default: false
    config_param :auto_delete, :bool, default: false
    config_param :passive, :bool, default: false
    config_param :payload_format, :string, default: "json"
    config_param :tag_key, :bool, default: false
    config_param :tag_header, :string, default: nil
    config_param :time_header, :string, default: nil
    config_param :tls, :bool, default: false
    config_param :tls_cert, :string, default: nil
    config_param :tls_key, :string, default: nil
    config_param :tls_ca_certificates, :array, default: nil
    config_param :tls_verify_peer, :bool, default: true
    config_param :bind_exchange, :bool, default: false
    config_param :exchange, :string, default: ""
    config_param :routing_key, :string, default: "#"                       # The routing key used to bind queue to exchange - # = matches all, * matches section (tag.*.info)

    def configure(conf)
      conf['format'] ||= conf['payload_format'] # legacy
      compat_parameters_convert(conf, :parser)

      super

      parser_config = conf.elements('parse').first
      if parser_config
        @parser = parser_create(conf: parser_config)
      end

      @conf = conf
      unless (@host || @hosts) && @queue
        raise Fluent::ConfigError, "'host(s)' and 'queue' must be all specified."
      end
      check_tls_configuration
    end

    def start
      super
      # Create a new connection, unless its already been provided to us
      @connection = Bunny.new get_connection_options unless @connection
      @connection.start
      @channel = @connection.create_channel
      q = @channel.queue(@queue, passive: @passive, durable: @durable,
                       exclusive: @exclusive, auto_delete: @auto_delete)
      if @bind_exchange
        log.info "Binding #{@queue} to #{@exchange}, :routing_key => #{@routing_key}"
        q.bind(exchange=@exchange, routing_key: @routing_key)
      end

      q.subscribe do |delivery, meta, msg|
        log.debug "Recieved message #{@msg}"
        payload = parse_payload(msg)
        router.emit(parse_tag(delivery, meta), parse_time(meta), payload)
      end
    end # AMQPInput#run

    def shutdown
      log.info "Closing connection"
      @connection.stop
      super
    end

    private
    def parse_payload(msg)
      if @parser
        parsed = nil
        @parser.parse msg do |_, payload|
          if payload.nil?
            log.warn "failed to parse #{msg}"
            parsed = { "message" => msg }
          else
            parsed = payload
          end
        end
        parsed
      else
        { "message" => msg }
      end
    end

    def parse_tag( delivery, meta )
      if @tag_key && delivery.routing_key != ''
        delivery.routing_key
      elsif @tag_header && meta[:headers][@tag_header]
        meta[:headers][@tag_header]
      else
        @tag
      end
    end

    def parse_time( meta )
      if @time_header && meta[:headers][@time_header]
        Fluent::EventTime.from_time(Time.parse( meta[:headers][@time_header] ))
      else
        Fluent::Engine.now
      end
    end

    def check_tls_configuration()
      if @tls
        unless @tls_key && @tls_cert
            raise Fluent::ConfigError, "'tls_key' and 'tls_cert' must be all specified if tls is enabled."
        end
      end
    end

    def get_connection_options()
      hosts = @hosts ||= Array.new(1, @host)
      opts = {
        hosts: hosts, port: @port, vhost: @vhost,
        pass: @pass, user: @user, ssl: @ssl,
        verify_ssl: @verify_ssl, heartbeat: @heartbeat,
        tls: @tls,
        tls_cert: @tls_cert,
        tls_key: @tls_key,
        verify_peer: @tls_verify_peer
      }
      opts[:tls_ca_certificates] = @tls_ca_certificates if @tls_ca_certificates
      return opts
    end

  end # class AMQPInput

end # module Fluent::Plugin
